### PR TITLE
feat: Entity migration CLI

### PR DESCRIPTION
**Pull request by joaoandre-avaiga**

### CODE REVIEW

File: src/taipy/_cli/_migrate_cli.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate_cli.py]: Instead of `--path`, how about `--repository-type` option with `nargs='*'` to parse all command-line arguments into a list.
- The first string is either "fs/filesystem, sql, mongodb"?
- After declaring the repo type:
  - if the repo type is fs, then the user can input a path (or default is ".data")
  - if the repo type is mongo/sql, then the user can input host, port, path, ...

The usage would be something like:
```
taipy migrate filesystem fspath/

taipy migrate sql localhost 8080 database_path.sqlite

taipy migrate mongodb localhost 8080 username password
```

File: src/taipy/_cli/_migrate_cli.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate_cli.py]: We need to parse the argument to somehow call to `_migrate_mongo_entities()` and `_migrate_sql_entities()` as well

File: src/taipy/_cli/_migrate.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate.py]: This code block is duplicated on all `_migrate_..._entities()`. Probably should split it to a dedicated method.
```suggestion
    __migrate_entities(entities)
```

File: src/taipy/_cli/_migrate.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate.py]: Can we add a docstring clarify which Taipy version "parent_id" has been removed by "parent_ids"?

Further step may be splitting the methods that migrate from Taipy 1.0 -> Taipy 2.0 to a dedicated file, Taipy 2.0 -> Taipy 3.0 to a dedicated file, etc.

Comment (joaoandre-avaiga) [reply src/taipy/_cli/_migrate.py]: I believe it was on 2.1. Theres no migration from taipy 1.0 from 2.0. Just 2.X to 3.0. If I recall correctly we had a migration from 1.0 to 2.0 in code. @jrobinAV What do you think about migrating entities from 1.0? It shouldnt be too much work and we could add it later as well, to not block the release.

File: src/taipy/_cli/_migrate_cli.py
Comment (joaoandre-avaiga) [reply src/taipy/_cli/_migrate_cli.py]: Great idea. I was thinking about a way to handle the different repositories. I will add your suggestion.

File: src/taipy/_cli/_migrate.py
Comment (jrobinAV) [reply src/taipy/_cli/_migrate.py]: I believe we don t need to migrate from 1.0.

File: src/taipy/_cli/_migrate_cli.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate_cli.py]:
```suggestion
            "--repository-type", nargs="+", help="The repository type and additional repository properties"
```
My apology, the nargs should be "+" since we require at least 1 parameter. Also, we need to make the help string clearer for the user

File: src/taipy/_cli/_migrate_cli.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate_cli.py]: we should check if the `args.repository_type[1]` exists or not first, if not then assign to ".data"

File: src/taipy/_cli/_migrate_cli.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate_cli.py]: we should check if the `args.repository_type[1]` exists or not first, if not then raise an `argparse.ArgumentError`

File: src/taipy/_cli/_migrate_cli.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate_cli.py]: we should check if the `args.repository_type[1]` exists or not first, if not then we pass nothing to the method to use default value

File: tests/cli/test_cli.py
Comment (jrobinAV) [inline tests/cli/test_cli.py]: Should we add a word saying it should be done after changing the code ?
Check if this is really the case.

**Review by jrobinAV — DISMISSED** [review]

File: src/taipy/_cli/_migrate.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate.py]:
```suggestion
    datanode["owner_id"] = datanode.get("parent_id", None)
```
This line should migrate "parent_id" to "owner_id". Can you check again?
Also, I believe we need to pop the parent_id if it exists after migrating to owner_id

File: src/taipy/_cli/_migrate.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate.py]: This should also check if all output data nodes has `cacheable == True` to set it to True.

File: src/taipy/_cli/_migrate.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate.py]: We need to make sure to use the `cacheable` to determine the value of task's `skippable` attribute first before popping it.

File: src/taipy/_cli/_migrate.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate.py]:
```suggestion
if last_edition_date := datanode.pop("last_edition_date", None):
    datanode["last_edit_date"] = last_edition_date
if edition_in_progress := datanode.pop("edition_in_progress", None):
    datanode["edit_in_progress"] = edition_in_progress

```
We also need to migrate the `last_edition_date` and `edition_in_progress` attributes.

File: src/taipy/_cli/_migrate.py
Comment (trgiangdo) [inline src/taipy/_cli/_migrate.py]:
```suggestion
        task["owner_id"] = task["parent_id"][0] if task["parent_id"] else None
```
I believe we need to migrate `parent_id`, not `parent_ids`, to `owner_id`.
Also, if the `parent_id` exists, we need to delete it.

**Review by trgiangdo — CHANGES_REQUESTED** [review]
There are quite a few migration method missing from the migration script:
- `_migrate_subscriber()` that rename scheduler to orchestrator on old jobs
- Migrate cacheable to skippable, mainly the part that if all outputs of a task is cacheable, then the task is skippable
- Migrate `last_edition_date` of a DataNode to `last_edit_date`
- Migrate `edition_in_progress` of a DataNode to `edit_in_progress`
- Migrate an entity that does not have the "version" field

File: src/taipy/_cli/_migrate.py
Comment (jrobinAV) [reply src/taipy/_cli/_migrate.py]: After discussion, we proposed to add migration from 1.0 in another PR that should not block the 3.0.0 release.

File: src/taipy/_cli/_migrate.py
Comment (jrobinAV) [reply src/taipy/_cli/_migrate.py]: well spotted.

File: src/taipy/_cli/_migrate.py
Comment (jrobinAV) [reply src/taipy/_cli/_migrate.py]: Is not this for 1.0 migration?

**Comment by joaoandre-avaiga** [comment]
The migrate cli was moved to core. This PR now contains just the entrypoint configuration for the CLI. It needs to be merged after merging the PR on taipy-core

**Review by jrobinAV — DISMISSED** [review]

**Review by trgiangdo — DISMISSED** [review]

**Review by jrobinAV — APPROVED** [review]

**Pull request merged**
